### ISSUE TITLE
How can I transfer learning of some specific classes from the pre-trained weights to my custom training data?

### ISSUE THREAD

**Issue by haquocviet** [AlexeyAB/darknet#7825]
Hello everyone,
I have some questions regarding training a custom dataset.
- Should I use any weights file, e.g. yolov4x-mish.conv.166 and yolov4x-mish.weights, for training my custom dataset? If the answer is no, what is the reason?
- How can I transfer learning of some specific classes from the pre-trained weights (e.g. COCO) to my custom training if my custom classes are different (adds some new ones and removes others) from the pre-trained classes?

Thanks.

**Comment by lsd1994** [comment]
1.You can use pretrain weights yolov4x-mish.conv.166 for training, or you can train from random weights.
2.Use pretrain weights for transfer learning.

**Comment by haquocviet** [comment]
@lsd1994 Thanks for your reply.
1. Could you clarify what is difference between the weights for inferrence and the weights for training? There must be a reason (I don't know) that AlexeyAB writes that to use pretrained weights such as xxx.conv.yyy for training.
2. You seem not understand my question. What I meant is I just want to transfer some classes of the pretrained weights to my new training data which has some new classes. Referenced to the answer of @stephanecharette [here](https://github.com/AlexeyAB/darknet/issues/7696#issuecomment-835867428), it seems there is no way to do so. That is reason why I closed this issue earlier.

**Comment by lsd1994** [comment]
> @lsd1994 Thanks for your reply.
> 
> 1. Could you clarify what is difference between the weights for inferrence and the weights for training? There must be a reason (I don't know) that AlexeyAB writes that to use pretrained weights such as xxx.conv.yyy for training.

yolov4x-mish.conv.166 is pretrain weights, only has weights of backbone(166 layers), used for training.
yolov4x-mish.weights is final weights, has all weights of the network, used for inference.

> 2. You seem not understand my question. What I meant is I just want to transfer some classes of the pretrained weights to my new training data which has some new classes. Referenced to the answer of @stephanecharette [here](https://github.com/AlexeyAB/darknet/issues/7696#issuecomment-835867428), it seems there is no way to do so. That is reason why I closed this issue earlier.

Yes there is no way to do that, you can only transfer some layers(or weights), rather than classes.

**Comment by haquocviet** [comment]
`Yes there is no way to do that, you can only transfer some layers(or weights), rather than classes.`
As I know the yolov4x-xxx.conv.166 is the pretrained weights of the COCO dataset which has 80 classeses, in order to make my custom weight to "learn" the ability of detection of such object classes, have I define my custom dataset to have same 80 classes?

**Comment by lsd1994** [comment]
No, any class num is ok.

**Comment by leonzchang** [comment]
check #2139
you also can train your own model with/without pre-trained weights.

**Comment by dejongyeong** [comment]
@haquocviet 

I have the same question here which I need further clarification - just started learning YOLO. 

I have other custom classes to train on top of the pre-trained weights (which are trained on COCO dataset). The questions are:

- while training, should I the class labels that needed on the .names file?
- also, should I include images and their corresponding bounding boxes as well apart from the custom objects? E.g., person (COCO dataset) and shipping containers (custom) in a port environment.

Thanks and looking forward to hearing back from you.

**Comment by haquocviet** [comment]
@dejongyeong Except "shipping contianers" that I don't know, the rest of your questions are some of necessary steps for training a custom dataset. For more details, you should read the instructions [here](https://github.com/AlexeyAB/darknet/wiki/Train-Detector-on-MS-COCO-(trainvalno5k-2014)-dataset).

**Comment by dejongyeong** [comment]
@haquocviet thanks for the information but not sure if you understand my question correctly here or I am just confusing. What I meant was if I need only specific classes from the COCO dataset and other classes of my own (custom terminal port-related objects), would that be possible to achieve when using pre-trained weights (transfer learning)?

- Should I include only the custom class labels in the ".names" file or all COCO labels including custom labels?

**Comment by haquocviet** [comment]
Actually, at the time I asked the question here, I gave up transfer learning because I did not find a way to match the classes' ID of my custom datasets that has a few classes to the ones of 80 COCO classes. Hence I just followed the instructions as I mentioned before to train my own datasets.

**Comment by dejongyeong** [comment]
To make it clear, in other words, you are using the `csdarknet53-omega.conv.105` weight file with the addition of your custom dataset for training purposes?

**Comment by haquocviet** [comment]
No, I used this one yolov4.conv.137.

**Comment by dejongyeong** [comment]
@haquocviet thanks for the information. Just another question here that I would like to seek clarification on, if I trained a custom dataset (no person class) using the pre-trained model, but wanted to detect a person as well during testing, would that be possible? or the model will now just detect only the custom classes? Thanks.

**Comment by haquocviet** [comment]
I have not tried that case but as I stated previously, the class ID of your own dataset must matched to class ID of pretrained dataset (COCO). So far I haven't figure out what correct way to do class ID matching is, but one thing I know is the class ID of people of COCO's objects list is zero, hence the objects' class ID of your dataset should start from 1.
